src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,684 @@
+        Objects.requireNonNull(provider, "provider must not be null");
+        try {
+            Instance instance = GetInstance.getInstance("KDF", KDFSpi.class,
+                                                        algorithm,
+                                                        kdfParameters,
+                                                        provider);
+            if (!JceSecurity.canUseProvider(instance.provider)) {
+                String msg = "JCE cannot authenticate the provider "
+                             + instance.provider.getName();
+                throw new SecurityException(msg);
+            }
+            return new KDF(new Delegate((KDFSpi) instance.impl,
+                                        instance.provider), algorithm, kdfParameters);
+
+        } catch (NoSuchAlgorithmException nsae) {
+            return handleException(nsae);
+        }
+    }
+
+    private static KDF handleException(NoSuchAlgorithmException e)
+        throws NoSuchAlgorithmException, InvalidAlgorithmParameterException {
+        Throwable cause = e.getCause();
+        if (cause instanceof InvalidAlgorithmParameterException) {
+            throw (InvalidAlgorithmParameterException) cause;
+        }
+        throw e;
+    }
+
+    /**
+     * Derives a key, returned as a {@code SecretKey} object.
+     *
+     * @param alg
+     *     the algorithm of the resultant {@code SecretKey} object
+     * @param derivationSpec
+     *     the object describing the inputs to the derivation function
+     *
+     * @return the derived key
+     *
+     * @throws InvalidAlgorithmParameterException
+     *     if the information contained within the {@code derivationSpec} is
+     *     invalid or if the combination of {@code alg} and the {@code derivationSpec}
+     *     results in something invalid
+     * @throws NoSuchAlgorithmException
+     *     if {@code alg} is empty or invalid
+     * @throws NullPointerException
+     *     if {@code alg} or {@code derivationSpec} is null
+     */
+    public SecretKey deriveKey(String alg,
+                               AlgorithmParameterSpec derivationSpec)
+        throws InvalidAlgorithmParameterException, NoSuchAlgorithmException {
+        if (alg == null) {
+            throw new NullPointerException(
+                "the algorithm for the SecretKey return value must not be null");
+        }
+        if (alg.isEmpty()) {
+            throw new NoSuchAlgorithmException(
+                "the algorithm for the SecretKey return value must not be "
+                + "empty");
+        }
+        Objects.requireNonNull(derivationSpec);
+        if (delegateAndSpiAreInitialized(pairOfSpiAndProv)) {
+            return pairOfSpiAndProv.spi().engineDeriveKey(alg, derivationSpec);
+        } else {
+            return (SecretKey) chooseProvider(alg, derivationSpec);
+        }
+    }
+
+    /**
+     * Obtains raw data from a key derivation function.
```

Comment:
The first sentences of the two `derive` methods use different verbs: `Derives` and `Obtains`. Is it possible to use a same one?